src/util/check.h

```diff
@@ -38,4 +39,59 @@ class NonFatalCheckError : public std::runtime_error
         }                                                         \
     } while (false)
 
+/**
+ * ASSUME(expression)
+ *
+ * ASSUME(...) is used to document assumptions.
+ *
+ * Bitcoin Core is always compiled with assertions enabled. assert(...)
+ * is thus only appropriate for documenting critical assumptions
+ * where a failed assumption should lead to immediate abortion also
+ * in production environments.
+ *
+ * ASSUME(...) works like assert(...) if -DABORT_ON_FAILED_ASSUME
+ * and is side-effect safe.
+ *
+ * More specifically:
+ *
+ * * If compiled with -DABORT_ON_FAILED_ASSUME (set by --enable-debug
+ *   and --enable-fuzz): Evaluate expression and abort if expression is
+ *   false.
+ *
+ * * Otherwise:
+ *   Evaluate expression and continue execution.
+ *
+ * ABORT_ON_FAILED_ASSUME should not be set in production environments.
+ *
+ * Example
+ * =======
+ * int main(void) {
+ *     ASSUME(IsFoo());
+ *     ...
+ * }
+ *
+ * If !IsFoo() and -DABORT_ON_FAILED_ASSUME, then:
+ *     filename.cpp:123: main: ASSUME(IsFoo()) failed.
+ *     Aborted
+ * Otherwise the execution continues.
+ */
+
+#ifdef ABORT_ON_FAILED_ASSUME
+#define ASSUME_FAILED(expression, file, line, func)                                                  \
+    do {                                                                                             \
+        tfm::format(std::cerr, "%s:%d: %s: ASSUME(%s) failed. You may report this issue here: %s\n", \
+            file, line, func, expression, PACKAGE_BUGREPORT);                                        \
```

Comment:
Shouldn't this be logged even if the debug flag is not set at compile time?

Reply:
That's not how I intended it to work, but now that you suggest it there is perhaps no downside in logging it also in non-debug/non-fuzz mode? Since we're only writing to stderr (and not to disk) it should be safe. I'll update.